### ISSUE TITLE
There is no a way to track changes performed by hooks

### ISSUE THREAD

**Issue by PhantomYdn** [orientechnologies/orientdb#4575]
Guys,

There is the following testcase expected to be OK:

``` java
@Test
    public void testDocumentTrackingComplex()
    {
        final String className = "TestDocumentTrackingComplex";
        ODatabaseDocument db = wicket.getTester().getDatabase();
        OSchema schema = db.getMetadata().getSchema();
        final OClass classA = schema.createClass(className);
        classA.createProperty("a", OType.STRING);
        classA.createProperty("b", OType.STRING);
        db.registerHook(new ODocumentHookAbstract() {

            {
                setIncludeClasses(className);
            }

            @Override
            public DISTRIBUTED_EXECUTION_MODE getDistributedExecutionMode() {
                return DISTRIBUTED_EXECUTION_MODE.SOURCE_NODE;
            }

            @Override
            public void onRecordAfterCreate(ODocument iDocument) {
                System.out.println("onRecordAfterCreate");
                iDocument.field("a", "onRecordAfterCreate");
            }

            @Override
            public void onRecordAfterUpdate(ODocument iDocument) {
                System.out.println("onRecordAfterUpdate");
                iDocument.field("a", "onRecordAfterUpdate");
            }

            @Override
            public RESULT onRecordBeforeUpdate(ODocument iDocument) {
                System.out.println("onRecordAfterUpdate before undo: "+iDocument.field("a"));
                iDocument.undo("a");
                System.out.println("onRecordAfterUpdate after  undo: "+iDocument.field("a"));
                return RESULT.RECORD_CHANGED;
            }

            @Override
            public void onRecordAfterRead(ODocument iDocument) {
                assertEquals("original", iDocument.field("a"));
            }
        });
        db.commit();
        ODocument doc = new ODocument(classA);
        doc.field("a", "original");
        doc.save();
        doc.field("b", "other change");
        doc.save();
        doc.reload();
        doc.field("a", "updated");
        doc.save();
        doc.field("b", "other change");
        doc.save();
        doc.reload();
        doc.delete();
    }
```

But it's failing.
Simply saying: hook is changing document after create/read, but it is trying to undo its changes in onUpdate. But it seems that all changes performed during create/read in hooks are not tracked: so can't be rolled backed.

P.S. I checked doc.isTrackingChanges() during lifecycle of a document with the following testcase and it's not failing. So tracking of changes is enabled on document during hooks invocation.

``` java
    @Test
    public void testDocumentTrackingSimple()
    {
        final String className = "TestDocumentTrackingSimple";
        ODatabaseDocument db = wicket.getTester().getDatabase();
        OSchema schema = db.getMetadata().getSchema();
        final OClass classA = schema.createClass(className);
        classA.createProperty("a", OType.STRING);
        db.registerHook(new ORecordHook() {

            @Override
            public void onUnregister() {
                // NOP
            }

            @Override
            public RESULT onTrigger(TYPE iType, ORecord iRecord) {
                if(iRecord instanceof ODocument) {
                    ODocument doc = (ODocument) iRecord;
                    if(classA.isSuperClassOf(doc.getSchemaClass())) {
                        System.out.println("During "+iType+" document should track changes: "+doc.isTrackingChanges());
                        assertTrue("During "+iType+" document should track changes", doc.isTrackingChanges());
                    }
                }
                return RESULT.RECORD_NOT_CHANGED;
            }

            @Override
            public DISTRIBUTED_EXECUTION_MODE getDistributedExecutionMode() {
                return DISTRIBUTED_EXECUTION_MODE.SOURCE_NODE;
            }

        });
        db.commit();
        ODocument doc = new ODocument(classA);
        doc.field("a", "test");
        doc.save();
        doc.reload();
        doc.field("a", "test2");
        doc.save();
        doc.reload();
        doc.delete();
    }
```

So, what do you think?

**Comment by PhantomYdn** [comment]
Guys, any thoughts on this?

**Comment by PhantomYdn** [comment]
Guys, any thoughts on this one?

**Comment by PhantomYdn** [comment]
Any thoughts about this one? Probably it's expected to be fixed in 2.2.0? Unfortunatly can't test by myself due to some bugs in 2.2.0

**Comment by PhantomYdn** [comment]
Dear OrientDB team, is this issue finally got fixed in some latest versions?
